### PR TITLE
C#: Postupdate notes for ternary expressions.

### PR DESCRIPTION
**Pull request by michaelnebel**
This is the C# version of: https://github.com/github/codeql/pull/10444
That is, we can now identify flow out through arguments where expressions are inlined.

Assume we have
```csharp
public class SimpleClass
{
    public string field = "";
}

public void TaintField(SimpleClass sc)
{
    sc.field = "taint source";
}
```
then flow out via `x` and `y` are tracked correctly in the following pseudo example.
```csharp
SimpleClass x = ...
SimpleClass y = ...
TaintField(b ? x : y);
Sink(x.field);
Sink(y.field);
``` 

Furthermore, we also track flow out of expressions like
```csharp
TaintField((SimpleClass)x) // Cast expressions
TaintField(x ?? y) // Null Coalescing expressions
TaintField(x!) // Suppress nullable warning expressions
TaintField(y = x) // Assignment expressions
TaintField(choice switch "x" => x ...)
```

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:185]:
```suggestion
  any(LocalExprStepConfiguration x).hasExprPath(_, result, any(ConditionalExpr ce), cfn)
```

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:185]: The reason for this change is that it respects CFG splitting.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:188]: Perhaps just inline?

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1946]: `m(b ? x : y)`

**Review by hvitved — CHANGES_REQUESTED** [review]

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:427]: Let's join these two `exists` into one.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:419]: We should also include
- `NullCoalescingExpr::getAnOperand`
- `SwitchExpr::getACase`
- `SwitchCaseExpr::getBody`

(And ideally also tests.)

**Review by hvitved — CHANGES_REQUESTED** [review]
We are missing some cases, otherwise LGTM.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:419]: Alright - I will look into it!

**Review by hvitved — APPROVED** [review]
Thanks, LGTM. Let's see what DCA says before merging.

**Comment by michaelnebel** [comment]
DCA execution has finished.
(1) (Unfortunately) no new alerts were found.
(2) There doesn't appear to be any performance regressions.

**Pull request merged**
